backend/src/main/scala/bloop/util/HashedSource.scala

```diff
@@ -0,0 +1,43 @@
+package bloop.util
+
+import java.io.ByteArrayInputStream
+import java.io.InputStream
+import java.nio.file.Path
+
+import bloop.reporter.Problem
+
+import sbt.internal.inc.HashUtil
+import sbt.internal.inc.PlainVirtualFileConverter
+import xsbti.BasicVirtualFileRef
+import xsbti.FileConverter
+import xsbti.PathBasedFile
+import xsbti.VirtualFile
+
+/**
+ * In memory VirtualFile that has a hash corresponding to the actual file
+ * being given to scala and java compielrs.
+ *
+ * @param content the content of the file
+ * @param bloopHash the hash of the file based on the input
+ * @param path path to the file
+ */
+case class HashedSource(content: Array[Byte], bloopHash: Int, path: Path)
+    extends BasicVirtualFileRef(path.toString)
+    with VirtualFile
+    with PathBasedFile {
+
+  override def contentHash(): Long = HashUtil.farmHash(content)
+
+  override def name(): String = path.getFileName.toString
+
+  override def toPath(): Path = path
+
+  override def input(): InputStream = {
+    new ByteArrayInputStream(content)
+  }
+  override def toString: String = s"VirtualSourceFile($Problem.id@${bloopHash})"
+}
+
+object HashedSource {
+  val converter: FileConverter = PlainVirtualFileConverter.converter
```

Comment:
This can be improved, because FileConverter is used to convert 
VirtualFileRef into VirtualFile, and HashedSource implement boths. Here you will probably create new instance on each conversion while you can just upcast

Reply:
Yeah, you can even go step further and
```
        case virtualFile: VirtualFile => virtualFile
```

Reply:
done